### DISCUSSION TITLE
Constructor-less Extension Class

### DISCUSSION THREAD

**Discussion by PaulTaro** [dotnet/csharplang#6963] [Language Ideas] [answered]
I would like to have a new language element "extension" in C#. It should be able to define all kinds of class elements except constructors. It should be able to be derived from another "extension" and implement interfaces. In short, I want an interface with implementation or a lightweight multiple inheritance.

This would be very handy if you want to derive several foreign classes and extend them with identical class members.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/issues/5497

**Reply by PaulTaro** [reply]
I find the proposal #5497 is very complex and not easy to understand. My wish was actually only to be able to do without T4 text templates in order to generate similar extensions for different types. With T4 templates you have to do without syntax-high-lighting and intellisense. With C# I can distribute the code of a type so wonderfully on several files. Why not also include files that have no explicit reference to the receiving type?

Here is an example:
```
extension XCell : ICell
{
	public nfloat Width
	{ get; set; }
	
	public nfloat Height
	{ get; set; }
	
	public Thickness  Margin
	{ get; set; }
	
	public Thickness  Padding
	{ get; set; }
	
	public CGSize DetermineContentSize(nfloat maxWidth, nfloat maxHeight)
	{
		…
	}
}

public class Cell : UIView, XCell
{
}
public class ScrollView : UIScrollView, XCell
{
}
public class TextBlock : UILabel, XCell 
{
}
```
and so on...

**Comment by TahirAhmadov** [comment, marked as answer]
What you are describing is called mixins. There are a few such discussions in this repository, including one I created.

**Reply by HaloFour** [reply]
Or traits, which you can sorta implement with default members on interfaces.

**Comment by HaloFour** [comment]
> In short, I want an interface with implementation or a lightweight multiple inheritance.

There's nothing lightweight about multiple inheritance.  Your example shows state, via auto-properties, which immediately makes multiple inheritance very hard as you have to deal with how to share or isolate that state with diamond problems.  It's one of the reasons the .NET doesn't offer multiple inheritance.

The closest you can get would be to use interfaces with default implementations.  You can include protected members to represent the state, but each implementing class would need to provide the storage, whether that be an auto-property or a field, etc.

**Comment by PaulTaro** [comment]
Default interface implementation does not help when I want to derive and extend foreign classes with identical properties.
So, I will stay with T4 templates or just copy and paste that stuff in multiple files with different (partial) class names.
